managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -2519,7 +2525,8 @@ public Position getPersistentMarkDeletedPosition() {
     public void rewind() {
         lock.writeLock().lock();
         try {
-            PositionImpl newReadPosition = ledger.getNextValidPosition(markDeletePosition);
+            PositionImpl newReadPosition = isCursorReadFromCompacted() ? markDeletePosition.getNext() :
+                    ledger.getNextValidPosition(markDeletePosition);
```

Comment:
If we don't know the compaction horizon here, will it introduce a new issue if the read compacted is enabled but the topic don't have any compacted data (All the messages are 
 not compacted)?

Reply:
I don't think this affects correctness, if the newReadPosition's ledger doesn't exist in ml, it will be moved to the beginning of the next ledger.

https://github.com/apache/pulsar/blob/b14fcb47f20398bdf5b698e158346dce19f0edc6/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L2074-L2085

Reply:
If so, can we always use `markDeletePosition.getNext()`? so that we don't need to check if the subscription is enabled read compacted or not.

Reply:
But this will affect the `CursorStats` and add some CPU cycles. I think we can apply this change only to cursors that have readCompacted enabled.

Reply:
@coderzc I see.

After checking the changes again. It looks like we don't need to persist the `readCompacted`?  If depends on the consumer settings. We only need to set the `readCompacted` when new consumer connected?

Reply:
Oh, I found a similar [PR](https://github.com/apache/pulsar/pull/13629) which fixed the Reader API.

Reply:
@codelipenghui 
> After checking the changes again. It looks like we don't need to persist the readCompacted? If depends on the consumer settings. We only need to set the readCompacted when new consumer connected?

I don't think this will work well because when recovering the cursor, maybe the consumer is not connected yet, then our md-position will be forced to move to the first position of the ml.
please see:
https://github.com/apache/pulsar/pull/21187/files#diff-fad355f91bd15cc041161f9a46fce62b7fee87fbfb8f0ff8a8b724a1bd1f29eeR685-R695

> Oh, I found a similar PR which fixed the Reader API.
 Yes, they are a issue

Reply:
@coderzc but it just moved the read position, right? If the first consumer connected and readCompacted is enabled, we can reset the read position to the next position of the MD position.

Or can we add a new method `rewind(boolean isReadCompacted)`?

Reply:
@codelipenghui  I mean we will lost the previous md-position after recovering the cursor if we don't record `readCompacted`.

Reply:
We can only skipping invalid md-position when recovering cursor and add `rewind(boolean readCompacted)` method.